### PR TITLE
add(selector): Add widget for selecting categorized blocks

### PR DESCRIPTION
**Pull request by BitScout**
I am targeting this branch, because this is a feature.

## Resolves
https://github.com/ekino/sonata/issues/20 Improve add block feature

## Changelog

```markdown
### Added
- Added (optional) replacement for "Add block" button, showing blocks in a categorized popup

### Fixed
- Fixed PHP-CS-Fixer config
```

## To do

- [x] Cleanup
- [x] Add tests

## Subject

Optionally replaces the "Add block" dropdown + button by a popup showing the blocks in in a categorized manner. This improves the usability in projects with lots of block types.

### CODE REVIEW

File: src/Controller/PageAdminController.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+     *
+     * @throws AccessDeniedException
+     * @throws NotFoundHttpException
+     *
+     * @return Response
+     */
+    public function composeContainerShowAction(Request $request = null)
+    {
+        /** @var BlockAdmin $blockAdmin */
+        $blockAdmin = $this->get('sonata.page.admin.block');
```
Comment (bmaziere) [inline src/Controller/PageAdminController.php]: we have to find a way to inject it. no more direct get from the container

File: src/Controller/PageAdminController.php [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        $id    = $request->get($this->admin->getIdParameter());
+        $block = $blockAdmin->getObject($id);
+
+        if (!$block) {
+            throw new NotFoundHttpException(sprintf('Unable to find the block with id : %s', $id));
+        }
+
+        /** @var BlockServiceManager $blockServiceManager */
+        $blockServiceManager = $this->get('sonata.block.manager');
```
Comment (bmaziere) [inline src/Controller/PageAdminController.php]: same

File: src/Controller/PageAdminController.php [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        $blockServices       = $blockServiceManager->getServicesByContext('sonata_page_bundle', false);
+
+        // Filter service using the template configuration
+        /** @var Page $page */
+        if ($page = $block->getPage()) {
+            /** @var TemplateManagerInterface $templateManager */
+            $templateManager = $this->get('sonata.page.template_manager');
```
Comment (bmaziere) [inline src/Controller/PageAdminController.php]: same

File: src/Controller/PageAdminController.php [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                }
+            }
+
+            // We're filtering remaining block services according to page
+            $blockServices = $this->blockFilter->filter($blockServices, $page);
+        }
+
+        $blocksByCategory = $this->getBlocksByCategory($blockServices);
+
+        return $this->renderWithExtraParams('@SonataHelpers/PageAdmin/compose_container_show.html.twig', [
```
Comment (bmaziere) [inline src/Controller/PageAdminController.php]: can you add a todo to explain that it is done like this for now but we will have to use template registry to get the template name in the future plz?

File: src/Controller/PageAdminController.php [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * Return blocks grouped by categories.
+     *
+     * @param array $blockServices
+     *
+     * @return array
+     */
+    protected function getBlocksByCategory(array $blockServices)
```
Comment (bmaziere) [inline src/Controller/PageAdminController.php]: `: array`

File: src/Block/BlockFilter/BlockFilter.php [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+use Sonata\PageBundle\Model\PageInterface;
+
+/**
+ * Class BlockFilter.
+ *
+ * This class can be used to filter blocks available on pages, by allowing or denying it according to the page
+ *
+ * @author Romain Mouillard <romain.mouillard@ekino.com>
+ */
+class BlockFilter
```
Comment (bmaziere) [inline src/Block/BlockFilter/BlockFilter.php]: should be final IMO

File: src/Block/BlockFilter/BlockFilter.php [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+{
+    /**
+     * @var array
+     */
+    protected $blockConfig;
```
Comment (bmaziere) [inline src/Block/BlockFilter/BlockFilter.php]: private for all

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * Class SonataAddBlockDialogPass.
+ *
+ * @author Christian Kollross <christian.kollross@ekino.com>
+ */
+class SonataAddBlockDialogPass implements CompilerPassInterface
```
Comment (bmaziere) [inline src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: can be final too

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public function process(ContainerBuilder $container)
```
Comment (bmaziere) [inline src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: phpdoc

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 10]
```diff
@@ -74,4 +75,17 @@ private function configureSonataMediaPrivateFileProvider(array $config, Containe
...
             ->replaceArgument(6, $config['allowed_mime_types']);
     }
+
+    /**
+     * @param array            $config
+     * @param ContainerBuilder $container
+     */
+    private function configureSonataAddBlockDialog(array $config, ContainerBuilder $container): void
+    {
+        $container
```
Comment (bmaziere) [inline src/DependencyInjection/SonataHelpersExtension.php]: should only be done if feature is enabled IMO

File: src/Resources/config/services.xml [hunk 11]
```diff
@@ -31,6 +31,16 @@
 
             <tag name="sonata.media.provider" />
         </service>
+
+        <service id="Sonata\HelpersBundle\Controller\PageAdminController">
+            <tag name="controller.service_arguments"/>
+            <argument type="service" id="Sonata\HelpersBundle\Block\BlockFilter\BlockFilter" />
+        </service>
+
+        <service id="Sonata\HelpersBundle\Block\BlockFilter\BlockFilter" public="true">
```
Comment (bmaziere) [inline src/Resources/config/services.xml]: no need for being public as it is injected in PAgeAdminController

File: src/SonataHelpersBundle.php [hunk 12]
```diff
@@ -22,4 +24,10 @@
  */
 class SonataHelpersBundle extends Bundle
 {
+    public function build(ContainerBuilder $container)
```
Comment (bmaziere) [inline src/SonataHelpersBundle.php]: phpdoc

**Comment by bmaziere** [comment]
your branch is out of date with the base branch. you should update your branch @BitScout

File: docs/40-categorized-add-block-dialog.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+```
+
+* `enabled`: Set to `false` to completely disable the feature. (Or do not add this configuration at all.)
+* `categories`: Key-label-list of all available categories. The empty key is the default category, assigned to all blocks without specified categories.
+* `block_config`: Assign one or more categories to each Block class.
+
+
+## Installation
+
+Install the node package `sortablejs`.
```
[the remark this answers is not in the document]
Comment (BitScout) [reply docs/40-categorized-add-block-dialog.md]: Removing this as `sortablejs` is no longer used.

File: src/Controller/PageAdminController.php (hunk 1 above)
Comment (BitScout) [reply src/Controller/PageAdminController.php]: Well, `<argument type="service" id="sonata.page.admin.block" />` doesn't work.

Comment (BitScout) [reply src/Controller/PageAdminController.php]: Sorry, my bad. It works if you assign the BlockAdmin in the constructor... :innocent:

File: docs/40-categorized-add-block-dialog.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+Edit your `webpack.config.js`:
+
```
[the remark this answers is not in the document]
Comment (BitScout) [reply docs/40-categorized-add-block-dialog.md]: Seems to work both with 2 and 3.

File: src/Resources/translations/SonataHelpersBundle.en.xliff [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+                <target>All</target>
+            </trans-unit>
+            <trans-unit id="generic.cancel">
+                <source>generic.cancel</source>
+                <target>Cancel</target>
+            </trans-unit>
+            <trans-unit id="block.category.default">
+                <source>block.category.default</source>
+                <target>Not categorized</target>
+            </trans-unit>
```
Comment (bmaziere) [inline src/Resources/translations/SonataHelpersBundle.en.xliff]: can you sort by trans-unit id plz?

File: src/Resources/translations/SonataHelpersBundle.fr.xliff [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+                <target>Tous</target>
+            </trans-unit>
+            <trans-unit id="generic.cancel">
+                <source>generic.cancel</source>
+                <target>Annuler</target>
+            </trans-unit>
+            <trans-unit id="block.category.default">
+                <source>block.category.default</source>
+                <target>Non catégorisé</target>
+            </trans-unit>
```
Comment (bmaziere) [inline src/Resources/translations/SonataHelpersBundle.fr.xliff]: same

File: src/Resources/views/PageAdmin/compose_container_show.html.twig [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        <button type="button" class="btn btn-default btn-sm" data-toggle="modal" data-target="#blockSelectModal">
+            <i class="fa fa-plus-circle"></i>
+            {{ 'admin.page.add_block'|trans }}
+        </button>
+        <div class="modal fade BlockSelectModal" id="blockSelectModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">
+            <div class="modal-dialog modal-lg" role="document">
+                <div class="modal-content">
+                    <div class="modal-header">
+                        <button type="button" class="close" data-dismiss="modal" aria-label="Fermer"><span aria-hidden="true">&times;</span></button>
+                        <h4 class="modal-title" id="myModalLabel">{{ 'admin.page.select_block'|trans({}) }}</h4>
```
Comment (bmaziere) [inline src/Resources/views/PageAdmin/compose_container_show.html.twig]: {{ 'admin.page.select_block'|trans }}

Comment (bmaziere) [reply src/Resources/views/PageAdmin/compose_container_show.html.twig]: for all in this file

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        $configs = $container->getExtensionConfig('sonata_helpers');
+        $enabled = $configs[0]['compose_container']['enabled'];
+
+        if($enabled && $container->hasParameter('sonata.page.admin.page.controller')) {
+            $container->setParameter('sonata.page.admin.page.controller', 'SonataHelpersBundle:PageAdmin');
+        }
+    }
+}
```
Comment (bmaziere) [inline src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: this class has to be unit tested

File: src/Block/BlockFilter/BlockFilter.php [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+            $blockCategories = $blockConfig['categories'] ?? [];
+
+            foreach ($blockCategories as $blockCategory) {
+                if (!isset($this->categories[$blockCategory])) {
+                    throw new \InvalidArgumentException(sprintf('Category "%s" defined for block "%s" is unknown.', $blockCategory, $code));
+                }
+            }
+        }
+    }
+}
```
Comment (bmaziere) [inline src/Block/BlockFilter/BlockFilter.php]: this class has to be unit tested

File: docs/40-categorized-add-block-dialog.md [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+  ...
+  path.resolve('/vendor/ekino/sonata/src/Resources/public/css/blockSelectModal.scss')
+]);
+...
+Encore.addEntry('...', [
+  ...
+  path.resolve('/vendor/ekino/sonata/src/Resources/public/js/MutationObserver.js')
+]);
+```
+
```
Comment (bmaziere) [inline docs/40-categorized-add-block-dialog.md]: we need to add something about the compiler pass here. We should explain that by enabling the feature, the PageAdminController will now be the one from this bundle and that it override `composeContainerShowAction` method from sonata page bundle one to make the feature available.

We could also add a word about that if you want this feature but you also need to override the page admin controller, then you should extend this PageAdminController

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /**
+     * @param ContainerBuilder $container
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $configs = $container->getExtensionConfig('sonata_helpers');
+        $enabled = $configs[0]['compose_container']['enabled'];
+
+        if($enabled && $container->hasParameter('sonata.page.admin.page.controller')) {
```
Comment (mremi) [inline src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: Wrong CS `if(`, should be detected by cs fixer

File: src/Block/BlockFilter/BlockFilter.php [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    private function checkBlockConfig(array $config): void
+    {
+        // Check block config validity
+        foreach ($config as $code => $blockConfig) {
+            $blockCategories = $blockConfig['categories'] ?? [];
+
+            foreach ($blockCategories as $blockCategory) {
+                if (!isset($this->categories[$blockCategory])) {
+                    throw new \InvalidArgumentException(sprintf('Category "%s" defined for block "%s" is unknown.', $blockCategory, $code));
```
Comment (mremi) [inline src/Block/BlockFilter/BlockFilter.php]: missing in phpdoc

File: src/Block/BlockFilter/BlockFilter.php [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+        return $blockCategories;
+    }
+
+    /**
+     * @return string|null
+     */
+    private function getDefaultCategory(): ?string
+    {
+        if (!isset($this->categories[$this->defaultCategory])) {
+            throw new \InvalidArgumentException(sprintf('Default category "%s" is not configured.', $this->defaultCategory));
```
Comment (mremi) [inline src/Block/BlockFilter/BlockFilter.php]: missing in phpdoc

File: src/Controller/PageAdminController.php [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        if (false === $this->blockAdmin->isGranted('LIST')) {
+            throw new AccessDeniedException();
+        }
+
+        $id    = $request->get($this->admin->getIdParameter());
+        $block = $this->blockAdmin->getObject($id);
+
+        if (!$block) {
+            throw new NotFoundHttpException(sprintf('Unable to find the block with id : %s', $id));
```
Comment (mremi) [inline src/Controller/PageAdminController.php]: missing in phpdoc

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 25]
```diff
@@ -74,4 +75,18 @@ private function configureSonataMediaPrivateFileProvider(array $config, Containe
...
             ->replaceArgument(6, $config['allowed_mime_types']);
     }
+
+    /**
+     * @param array            $config
+     * @param ContainerBuilder $container
+     */
+    private function configureSonataAddBlockDialog(array $config, ContainerBuilder $container): void
+    {
+        if($config['enabled']) {
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: CS fail here too

File: tests/DependencyInjection/Compiler/SonataAddBlockDialogPassTest.php [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+ * @author Christian Kollross <christian.kollross@ekino.com>
+ */
+class SonataAddBlockDialogPassTest extends TestCase
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function setUp(): void
+    {
+    }
```
Comment (bmaziere) [inline tests/DependencyInjection/Compiler/SonataAddBlockDialogPassTest.php]: seems useless no?

File: src/Controller/PageAdminController.php [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+     *
+     * @return Response
+     */
+    public function composeContainerShowAction(Request $request = null): Response
+    {
+        if (false === $this->blockAdmin->isGranted('LIST')) {
+            throw new AccessDeniedException();
+        }
+
+        $id    = $request->get($this->admin->getIdParameter());
```
Comment (mremi) [inline src/Controller/PageAdminController.php]: If data from query parameters for instance, better to use `$request->query->get`

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /**
+     * @param ContainerBuilder $container
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $configs = $container->getExtensionConfig('sonata_helpers');
+        $enabled = $configs[0]['compose_container']['enabled'];
+
+        if ($enabled && $container->hasParameter('sonata.page.admin.page.controller')) {
```
Comment (mremi) [inline src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: Why doing that here, in a compiler pass? Why don't do it in the extension?

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 29]
```diff
@@ -74,4 +75,18 @@ private function configureSonataMediaPrivateFileProvider(array $config, Containe
...
+
+    /**
+     * @param array            $config
+     * @param ContainerBuilder $container
+     */
+    private function configureSonataAddBlockDialog(array $config, ContainerBuilder $container): void
+    {
+        if ($config['enabled']) {
+            $container
+                ->findDefinition('Sonata\HelpersBundle\Block\BlockFilter\BlockFilter')
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: please use `::class`

File: tests/Block/BlockFilter/BlockFilterTest.php [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+ * Class BlockFilterTest.
+ *
+ * @author Christian Kollross <christian.kollross@ekino.com>
+ */
+class BlockFilterTest extends TestCase
+{
+    /**
+     * @var PageInterface
+     */
+    private $pageInterface;
```
Comment (mremi) [inline tests/Block/BlockFilter/BlockFilterTest.php]: should be `$page`

File: tests/DependencyInjection/SonataHelpersExtensionTest.php [hunk 31]
```diff
@@ -66,12 +67,29 @@ public function testSonataMediaConfig(): void
...
-            'allowed_mime_types' => ['baz'],
-        ]]], $container);
+        $blockFilterDefinition = $this->createMockDefinition();
+        $blockFilterDefinition->expects($this->exactly(2))->method('replaceArgument')
+            ->withConsecutive(
+                [0, ['alpha' => 'bravo']],
+                [1, ['charlie' => ['categories' => [], 'only_pages' => []]]]
+            )->willReturnSelf();
+        $container->expects($this->at(3))->method('findDefinition')
+            ->with('Sonata\HelpersBundle\Block\BlockFilter\BlockFilter')
```
Comment (mremi) [inline tests/DependencyInjection/SonataHelpersExtensionTest.php]: `::class`

File: .php_cs [hunk 32]
```diff
@@ -21,36 +21,19 @@ EOF;
...
-        'indentation_type' => true,
-        'linebreak_after_opening_tag' => true,
+        'header_comment' => array('header' => $header),
         'lowercase_constants' => true,
         'lowercase_keywords' => true,
         'method_separation' => true,
-        'native_function_invocation' => ['include' => ['@compiler_optimized']],
-        'no_alias_functions' => true,
         'no_closing_tag' => true,
         'no_extra_consecutive_blank_lines' => [
```
Comment (mremi) [inline .php_cs]: Why overriding Symfony rules?

Comment (BitScout) [reply .php_cs]: As far as I know `no_extra_consecutive_blank_lines` is not part of `@Symfony`...

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php (hunk 28 above)
Comment (BitScout) [reply src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: We only want to replace the `PageAdminController` if the feature is enabled. Is there another way for that?

File: .php_cs (hunk 32 above)
Comment (mremi) [reply .php_cs]: 1. `no_extra_consecutive_blank_lines` is deprecated, replaced by `no_extra_blank_lines `
2. see https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.15/src/RuleSet.php#L90

We should probably fix it in an another PR

File: src/Controller/PageAdminController.php (hunk 27 above)
Comment (BitScout) [reply src/Controller/PageAdminController.php]: `$request->query->get($this->admin->getIdParameter())` gives me `null` instead of the ID...

Comment (bmaziere) [reply src/Controller/PageAdminController.php]: dump the request ;) i'm sure you will find your id ;)

Comment (mremi) [reply src/Controller/PageAdminController.php]: I give you an example, where your data from?

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php (hunk 28 above)
Comment (mremi) [reply src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: In the extension

File: src/Controller/PageAdminController.php (hunk 27 above)
Comment (BitScout) [reply src/Controller/PageAdminController.php]: `$request->attributes->get($this->admin->getIdParameter());` would work...
image

File: src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php (hunk 28 above)
Comment (BitScout) [reply src/DependencyInjection/Compiler/SonataAddBlockDialogPass.php]: Cool, that works, and it saves us two classes.

**Comment by bmaziere** [comment]
> can update the changelog plz?

still relevant @BitScout ;)

File: src/DependencyInjection/Configuration.php [hunk 33]
```diff
@@ -42,6 +42,21 @@ public function getConfigTreeBuilder()
                         ->arrayNode('allowed_mime_types')->cannotBeEmpty()->prototype('scalar')->end()->defaultValue([])->end()
                     ->end()
                 ->end()
+                ->arrayNode('compose_container')
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->booleanNode('enabled')->defaultValue(false)->end()
```
Comment (mremi) [inline src/DependencyInjection/Configuration.php]: `defaultFalse()`

File: src/DependencyInjection/SonataHelpersExtension.php [hunk 34]
```diff
@@ -74,4 +76,22 @@ private function configureSonataMediaPrivateFileProvider(array $config, Containe
...
             ->replaceArgument(6, $config['allowed_mime_types']);
     }
+
+    /**
+     * @param array            $config
+     * @param ContainerBuilder $container
+     */
+    private function configureSonataAddBlockDialog(array $config, ContainerBuilder $container): void
+    {
+        if ($config['enabled']) {
```
Comment (mremi) [inline src/DependencyInjection/SonataHelpersExtension.php]: please add early return, I mean `if (!$config['enabled']) {`

File: src/Resources/translations/SonataHelpersBundle.en.xliff [hunk 35]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mremi) [inline src/Resources/translations/SonataHelpersBundle.en.xliff]: Could you please create the DE version? :)

**Pull request merged**